File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public void reverse(int[] output, int left, int right){
        while(left < right){
            int temp = output[left];
            output[left] = output[right];
            output[right] = temp;
            
            left++;
            right--;
        }
    }
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> bucket = new Stack<>();
        int n = temperatures.length;
        int output[] = new int[n];
        int j=0;
        
        for(int i=n-1; i>=0; i--){
            if(bucket.isEmpty()){
                output[j++] = 0;
                bucket.add(i);
            }
            else{
                while(!bucket.isEmpty() && temperatures[bucket.peek()] <= temperatures[i]){
                    bucket.pop();
                }
                if(bucket.isEmpty()){
                    output[j++] = 0;
                    bucket.add(i);
                }
                else{
                    output[j++] = bucket.peek() - i;
                    bucket.add(i);
                }
            }
        }
        
        reverse(output, 0, n-1);
        
        return output;
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        //store index of element in stack
        stack<int> bucket;
        vector<int> result;
        int n = temperatures.size();
        for(int i=n-1; i>=0; i--){
            if(bucket.empty()){
                result.push_back(0);
                bucket.push(i);
            }
            else{
                while(!bucket.empty() && temperatures[bucket.top()] <= temperatures[i]){
                    bucket.pop();
                }
                if(bucket.empty()){
                    result.push_back(0);
                    bucket.push(i);
                }
                else{
                    result.push_back(bucket.top() - i);
                    bucket.push(i);
                }
            }
        }
        reverse(result.begin(), result.end());
        return result;
    }
};